docker-compose.yml

```diff
@@ -0,0 +1,210 @@
+      - "--port=8181"
+      - "--admin_host=0.0.0.0"
+      - "--admin_port=8000"
+      - "--skip_sdk_update_check=True"
+
+  datastore:
+    build:
+      context: .
+      dockerfile: ./docker/Dockerfile
+    container_name: oppia-cloud-datastore
+    ports:
+      - "8089:8089"
+    environment:
+      - CLOUDSDK_CORE_DISABLE_PROMPTS=1
+    depends_on:
+      angular-build:
+        condition: service_healthy
+      webpack-compiler:
+        condition: service_healthy
+    healthcheck:
+      test: curl -f http://localhost:8089/ || exit 1
+      timeout: 10s
+      interval: 2m
+      retries: 5
+      start_period: 20m
+    entrypoint: /google-cloud-sdk/bin/gcloud
+    command:
+      - "beta"
+      - "emulators"
+      - "datastore"
+      - "start"
+      - "--project"
+      - "dev-project-id"
+      - "--data-dir"
+      - "/app/cloud_datastore_emulator_cache"
+      - "--host-port=0.0.0.0:8089"
+      - "--consistency=1.0"
+      - "--quiet"
+      - "--no-store-on-disk"
+
+  firebase:
+    image: node@sha256:48650ae5c5dc9fbd0d0559fd1248e779f7214ba99434420a089bc7853aeefb65
+    container_name: oppia-firebase-emulator
+    ports:
+      - "9099:9099"
+      - "4000:4000"
+    volumes:
+      - ./.firebase.json:/app/oppia/.firebase.json
+      - shared_node_modules:/app/oppia/node_modules
+    environment:
+      - oppia_is_dockerized=true
+    healthcheck:
+      test: wget --spider http://localhost:4000/auth/ && echo 'y' | /app/oppia/node_modules/firebase-tools/lib/bin/firebase.js emulators:export /app/firebase_emulator_cache --project=dev-project-id || exit 1
+      timeout: 20s
+      interval: 20s
+      retries: 5
+    entrypoint:
+      - /app/oppia/node_modules/firebase-tools/lib/bin/firebase.js
+      - "emulators:start"
+      - "--only"
+      - "auth"
+      - "--project=dev-project-id"
+      - "--config=/app/oppia/.firebase.json"
+    command:
+      - "--import=/app/firebase_emulator_cache"
+      - "--export-on-exit=/app/firebase_emulator_cache"
+
+  redis:
+    image: redis@sha256:aebc052b5a7f50d79f00a4deee22109e2ecf22f439140a03f2c9770055ffec7b
+    container_name: oppia-redis
+    ports:
+      - "6379:6379"
+    environment:
+      - PATH=/usr/local/bin/
+    volumes:
+      - oppia_redis_dump:/app/oppia/dump.db
+      - ./redis_docker.conf:/app/oppia/redis_docker.conf
+    healthcheck:
+      test: redis-cli ping
+      timeout: 5s
+      interval: 30s
+      retries: 5
+    command: redis-server /app/oppia/redis_docker.conf
+
+  elasticsearch:
+    image: elasticsearch@sha256:fa7141154a7e14df214e42f08c333702403eb88c02ba44e79322a1f42d733013
+    container_name: oppia-elasticsearch
+    ports:
+      - "9200:9200"
+    environment:
+      - ES_JAVA_OPTS=-Xms100m -Xmx500m
+      - ES_PATH_CONF=/usr/share/elasticsearch/config
+      - discovery.type=single-node
+    depends_on:
+      angular-build:
+        condition: service_healthy
+      webpack-compiler:
+        condition: service_healthy
```

Comment:
the depends_on used within the services donot define any meaning or relation I have added..
the reason is to have an optimised docker setup that takes least possible time to start. if we start all the containers at the same time, then there might be some resources issues as all the services are started in parallel. so I decided to have 2 stages in which services will start (this seems the most optimised way. please add your suggestions to improve more). so in stage1, webpack and angular-build (the most resource-taking services) are started, then in the stage 2, when the webpack and ng-build are completed with their build, rest other services are started.

Reply:
>  then there might be some resources issues as all the services are started in parallel. 

Considering you are having a low resource, have you seen any issue right now on running redis, ES and firebase together? If not, let's not add depends_on for now, once we will start facing issue after adding more services i.e, webpack, angular, we can think about parallelizing. (In general, we should use depends_on because some services depends on other)

>  so in stage1, webpack and angular-build (the most resource-taking services) are started, then in the stage 2, when the webpack and ng-build are completed with their build, rest other services are started.

Considering all the resources are going to run in parallel I'm not sure starting one after another will be much helpful? 

Maybe we can discuss this once we have all the services ready, we can fix it later!

Reply:
> Considering all the resources are going to run in parallel I'm not sure starting one after another will be much helpful?

ng-build and webpack takes little long for the first time when we start the containers, so we can have these 2 services started in the stage1, and then everything else in stage 2. I think this will be the most optimised way so far

Reply:
I would suggest focusing on optimizing each service for now and thinking about parallelization once we have all the services ready.

Reply:
aah, I still prefer to have this way to be followed. as the ng-build and webpack takes considerable amount of time for the first time